packages/@aws-cdk/aws-sagemaker/lib/model.ts

```diff
@@ -0,0 +1,405 @@
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import * as cdk from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { ContainerImage } from './container-image';
+import { ModelData } from './model-data';
+import { CfnModel } from './sagemaker.generated';
+
+/**
+ * Name tag constant
+ */
+const NAME_TAG: string = 'Name';
+
+/**
+ * Interface that defines a Model resource.
+ */
+export interface IModel extends cdk.IResource, iam.IGrantable, ec2.IConnectable {
+  /**
+   * Returns the ARN of this model.
+   *
+   * @attribute
+   */
+  readonly modelArn: string;
+
+  /**
+   * Returns the name of this model.
+   *
+   * @attribute
+   */
+  readonly modelName: string;
+
+  /**
+   * The IAM role associated with this Model.
+   */
+  readonly role?: iam.IRole;
+
+  /**
+   * Adds a statement to the IAM role assumed by the instance.
+   */
+  addToRolePolicy(statement: iam.PolicyStatement): void;
+}
+
+/**
+ * Represents a Model resource defined outside this stack.
+ */
+export interface ModelAttributes {
+  /**
+   * The name of this model.
+   */
+  readonly modelName: string;
```

Comment:
why `modelName` here? The "real" required property is `modelArn`, which you are then splicing together in the `Import` construct. However, this then requires the model to exist in the same environment that the stack is in. We're not serving customers who have models in different environments that they want to reference here.

I think this is what we want:

1) change `modelName` to `modelArn` here as a "required" attribute.
2) the implementation of `Model.fromModelName()` will need to do the arn generation before passing the arn into `fromModelAttributes`.
3) optionally include a `fromModelArn()` method.

In the actual implementation of the `Import` construct, I'll write a comment on how we can get the environment from the arn to support this use case i'm describing.

Reply:
By 'real', I mean that in the `fromModelAttributes` method further below, we do the following:

```ts
        this.modelArn = cdk.Stack.of(this).formatArn({
          service: 'sagemaker',
          resource: 'model',
          resourceName: this.modelName,
        });
```

i.e. we need the modelArn anyway.

While `Model` resources cannot be shared cross-account or cross-region today, if that were ever possible we would immediately have a bug in how we import resources. See https://github.com/aws/aws-cdk/pull/22549/files#r1008209404 for what that bug is.

I don't have a problem with having a `fromModelName` method the way you have it currently. I think we should also have a `fromModelArn` method. It would look something like this:

```ts
  public static fromModelArn(scope: Construct, id: string, arn: string): IModel {
    const modelName = // splice arn to get name
    return Model.fromModelAttributes(scope, id, { modelName, modelArn: arn });
  }

  public static fromModelName(scope: Construct, id: string, name: string): IModel {
    const modelArn = cdk.Stack.of(this).formatArn({
          service: 'sagemaker',
          resource: 'model',
          resourceName: this.modelName,
        });
    return Model.fromModelAttributes(scope, id, { modelName: name, modelArn: modelArn });
  }

  public static fromModelAttributes(scope: Construct, id: string, attrs: ModelAttributes): IModel {
    const modelName = attrs.modelName;
    const role = attrs.role;
    const modelArn = attrs.modelArn;

    class Import extends ModelBase {
      public readonly modelName = modelName;
      public readonly role = role;
      public readonly grantPrincipal: iam.IPrincipal;
      public readonly modelArn = modelArn;

      constructor(s: Construct, i: string) {
        super(s, i, {
          environmentFromArn: modelArn,
        });
        ...
      }
  }
```

The payoff is the `environmentFromArn`. Yes, it may be useless today because `Models` cannot be shared cross env. However, we are trying to standardize how we do imports for L2 to guard against this cross env sharing if/when it gets introduced.

Reply:
Does this mean I'd end up using a [technique like this](https://github.com/aws/aws-cdk/blob/cea1039e3664fdfa89c6f00cdaeb1a0185a12678/packages/%40aws-cdk/aws-events-targets/lib/util.ts#L114-L116) to prevent customers from attempting to use `IModel` across accounts-or-regions downstream within the `EndpointConfig` implementation (and from `IEndpointConfig` within the `Endpoint` implementation)?

Reply:
I think we can just do that in `fromModelAttributes` for now -- if the arn region / account is not the same as the scope region / account, then throw an error.

This is still better than before because we now explicitly throw an error in these scenarios, and can easily remove the synth-time check in the future if sagemaker decides to support this.